### PR TITLE
chore(live-network): only retry network for status code that makes sense

### PR DESCRIPTION
**Pull request by gre**
### ✅ Checklist

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - live-network

### 📝 Description

This is a generalisation of existing code to explicit the list of statusCodes that are allowed to be retried.

this avoid our stack to do 3 failing http code in a row for http code that will always fails anyway, for instance a Unauthorized case (where you fail authentication). 
**in live-hub context, this is a problem for Wallet Sync feature:**
Capture d’écran 2024-07-15 à 14 50 22

the committed statusCodes list is taken from https://github.com/sindresorhus/got/blob/main/documentation/7-retry.md#retry

### ❓ Context

- **JIRA or GitHub link**: na

---

### 🧐 Checklist for the PR Reviewers

- **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- **Any new dependencies** have been justified and documented.
- **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

### CODE REVIEW

**Review by valpinkman — DISMISSED** [review]

**Review by mcayuelas-ledger — DISMISSED** [review]

**Review by Justkant — APPROVED** [review]

File: libs/live-network/src/network.ts [hunk 1]
```diff
@@ -175,6 +175,10 @@ export type LiveNetworkResponse<T> = {
   data: T;
   status: number;
 };
+
+// inspired by https://github.com/sindresorhus/got/blob/main/documentation/7-retry.md#retry
+const retryableHttpStatusCodes = [408, 413, 429, 500, 502, 503, 504, 521, 522, 524];
```
Comment (sprohaszka-ledger) [inline libs/live-network/src/network.ts:180]: I find weird that we retry on server errors. Maybe "internal server" error, but "gateway" issue seems difficult to overcome.
Same for 413. If the payload is too huge, I am not sure it is a good idea to retry.

File: libs/live-network/src/network.ts [hunk 2]
```diff
@@ -198,9 +202,8 @@ export const newImplementation = async <T = unknown, U = unknown>(
       maxRetry: getEnv("GET_CALLS_RETRY"),
       retryCondition: error => {
         if (error && error.status) {
-          // A 422 shouldn't be retried without change as explained in this documentation
-          // https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422
-          return error.status !== 422;
+          // not all status codes are retryable
+          return retryableHttpStatusCodes.includes(error.status);
```
Comment (sprohaszka-ledger) [inline libs/live-network/src/network.ts:206]: Be careful, everyone has not migrate to this call.
You might need to duplicate the code with `implementation` function.

**Review by sprohaszka-ledger — COMMENTED** [review]
Not sure if we can rely on all those return codes.

File: libs/live-network/src/network.ts (hunk 1 above)
Comment (gre) [reply libs/live-network/src/network.ts:180]: i just have removed a lot of statusCodes, today's behavior is that they are ALL retried, so it's already a progress
but i agree with you tho, an ultimately target should to not even have this retry behavior on the network stack side imo

File: libs/live-network/src/network.ts (hunk 2 above)
Comment (gre) [reply libs/live-network/src/network.ts:206]: in my context i only need it fixed for `live-wallet`'s wallet sync code so at least it fixes this.
the benefits being that it limits the impact of the PR changes. only the new implementation users will benefit the evolution and i didn't break the behavior of the deprecated way.
when all users of live-network migrate to newImplementation they will then have the opportunity to verify this change is working 🤔

**Pull request merged**